File: backends/java/zappy-application/src/main/java/com/zappymart/application/accounts/AddToWishlist.java
package com.zappymart.application.accounts;

import com.zappymart.application.Visitor;
import com.zappymart.application.ports.ProductRepository;
import com.zappymart.application.ports.UnitOfWork;
import com.zappymart.application.ports.WishlistRepository;
import com.zappymart.domain.accounts.Wishlist;
import com.zappymart.domain.catalogue.Product;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserErrorCode;

import java.util.List;

public final class AddToWishlist {

    private final UnitOfWork unitOfWork;
    private final WishlistRepository wishlistRepository;
    private final ProductRepository productRepository;
    private final ViewWishlist viewWishlist;

    public AddToWishlist(UnitOfWork unitOfWork, WishlistRepository wishlistRepository,
                         ProductRepository productRepository, ViewWishlist viewWishlist) {
        this.unitOfWork = unitOfWork;
        this.wishlistRepository = wishlistRepository;
        this.productRepository = productRepository;
        this.viewWishlist = viewWishlist;
    }

    public Result<List<Product>> execute(Visitor visitor, String productId) {
        String ownerId = WishlistOwner.of(visitor);
        return unitOfWork.inTransaction(() -> {
            if (productRepository.byId(productId).isEmpty()) {
                return Result.refuse(UserErrorCode.PRODUCT_NOT_FOUND,
                        "No product has the id " + productId + ".", "productId");
            }
            Wishlist saved = wishlistRepository.save(wishlistRepository.ofOwner(ownerId).wishing(productId));
            return Result.of(viewWishlist.productsOf(saved));
        });
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/accounts/AddToWishlist.kt
package nl.zappymart.application.accounts

import nl.zappymart.application.Visitor
import nl.zappymart.application.ports.ProductRepository
import nl.zappymart.application.ports.UnitOfWork
import nl.zappymart.application.ports.WishlistRepository
import nl.zappymart.domain.shared.UserError
import nl.zappymart.domain.shared.UserErrorCode

class AddToWishlist(
    private val wishlists: WishlistRepository,
    private val products: ProductRepository,
    private val viewWishlist: ViewWishlist,
    private val wishlistOwner: WishlistOwner,
    private val unitOfWork: UnitOfWork,
) {

    fun execute(visitor: Visitor, productId: String): WishlistChange = unitOfWork.execute {
        val ownerId = wishlistOwner.forWriting(visitor)
        val anonymousCartId = if (visitor.isSignedIn) null else ownerId
        if (products.findById(productId) == null) {
            WishlistChange(
                viewWishlist.inWishlistOrder(wishlists.findByOwnerId(ownerId).productIds),
                anonymousCartId,
                listOf(UserError(UserErrorCode.PRODUCT_NOT_FOUND, "There is no product $productId.", "productId")),
            )
        } else {
            val saved = wishlists.findByOwnerId(ownerId).with(productId)
            wishlists.save(saved)
            WishlistChange(viewWishlist.inWishlistOrder(saved.productIds), anonymousCartId, emptyList())
        }
    }
}
